midend/predication.cpp

```diff
@@ -134,8 +134,9 @@ const IR::Node* Predication::clone(const IR::AssignmentStatement* statement) {
 }
 
 const IR::Node* Predication::preorder(IR::AssignmentStatement* statement) {
-    if (!inside_action || ifNestingLevel == 0)
+    if (!inside_action || ifNestingLevel == 0) {
```

Comment:
The field inside_action can be replaced by `findContext<IR::P4Action>() != null`. This will make the code easier to read and more robust too.

I would suggest commenting all data structures of the visitor as well, since your code structure has become quite complicated.